sync/class.jetpack-sync-client.php

```diff
@@ -0,0 +1,226 @@
+		add_action( 'spammed_comment', $handler, 10 );
+
+		// even though it's messy, we implement these hooks because 
+		// the edit_comment hook doesn't include the data
+		// so this saves us a DB read for every comment event
+		foreach ( array( '', 'trackback', 'pingback' ) as $comment_type ) {
+			foreach ( array( 'unapproved', 'approved' ) as $comment_status ) {
+				add_action( "comment_{$comment_status}_{$comment_type}", $handler, 10, 2 );
+			}
+		}
+
+		// options
+		add_action( 'added_option', $handler, 10, 2 );
+		add_action( 'updated_option', $handler, 10, 3 );
+		add_action( 'deleted_option', $handler, 10, 1 );
+
+		// themes
+		add_action( 'jetpack_sync_current_theme_support', $handler, 10 ); // custom hook, see meta-hooks below
+
+		// post-meta, and in the future - other meta?
+		foreach ( $this->meta_types as $meta_type ) {
+			// we need to make sure we don't commit before we receive these,
+			// because they're invoked after meta changes are saved to the DB
+			add_action( "added_{$meta_type}_meta", $handler, 99, 4 );
+			add_action( "updated_{$meta_type}_meta", $handler, 99, 4 );
+			add_action( "deleted_{$meta_type}_meta", $handler, 99, 4 );
+		}
+
+		/**
+		 * Other hooks - fire synthetic hooks for all the properties we need to sync,
+		 * e.g. when a theme changes
+		 */
+
+		// themes
+		add_action( 'switch_theme', array( $this, 'switch_theme_handler' ) );
+	}
+
+	function set_options_whitelist( $options ) {
+		$this->options_whitelist = $options;
+	}
+
+	function set_constants_whitelist( $constants ) {
+		$this->constants_whitelist = $constants;
+	}
+
+	function is_whitelisted_option( $option ) {
+		foreach ( $this->options_whitelist as $whitelisted_option ) {
+			if ( $whitelisted_option[0] === '/' && preg_match( $whitelisted_option, $option ) ) {
+				return true;
+			} elseif ( $whitelisted_option === $option ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	function set_codec( iJetpack_Sync_Codec $codec ) {
+		$this->codec = $codec;
+	}
+
+	function set_sync_queue( $queue ) {
+		$this->sync_queue = $queue;
+	}
+
+	function action_handler() {
+		$current_filter = current_filter();
+		$args           = func_get_args();
+
+		if ( $current_filter === 'wp_insert_post' && $args[1]->post_type === 'revision' ) {
+			return;
+		}
+
+		if ( in_array( $current_filter, array( 'deleted_option', 'added_option', 'updated_option' ) )
+		     &&
+		     ! $this->is_whitelisted_option( $args[0] )
+		) {
+			return;
+		}
+		Jetpack_Sync::schedule_sync();
+		$this->sync_queue->add( array(
+			$current_filter,
+			$args
+		) );
+	}
+
+	function switch_theme_handler() {
+		global $_wp_theme_features;
+
+		do_action( 'jetpack_sync_current_theme_support', $_wp_theme_features );
+	}
+
+	function do_sync() {
+		$this->maybe_sync_constants();
+
+		// TODO: only send buffer once, then do the rest in a cron job
+		$iters = 0;
+		while ( ( $buffer = $this->sync_queue->checkout() ) && $iters < 100 ) {
+
+			if ( ! $buffer ) {
```

Comment:
@gravityrail isn't this redundant since we have  `$buffer = $this->sync_queue->checkout()` on the loop check?